packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -68,8 +75,43 @@ export type DomAction =
       node: appDom.AppDomNode;
     };
 
+const undoStack: appDom.AppDom[] = [];
+const redoStack: appDom.AppDom[] = [];
+
+const updateUndoStack = throttle((dom: appDom.AppDom) => {
+  undoStack.push(dom);
+}, 1000);
+
+const SKIP_UNDO_ACTIONS = ['DOM_UNDO', 'DOM_REDO', 'DOM_SAVED', 'DOM_SAVING', 'DOM_SAVING_ERROR'];
+
 export function domReducer(dom: appDom.AppDom, action: DomAction): appDom.AppDom {
+  if (!SKIP_UNDO_ACTIONS.includes(action.type)) {
+    updateUndoStack(dom);
+  }
+
   switch (action.type) {
+    case 'DOM_UNDO': {
```

Comment:
If:
- user makes a change to a page
- user goes to custom component and makes a change
- user goes back to the page and makes another change
- user presses undo 3 times

If I understand correctly, this will undo the change to the component as well without any feedback to the user. How would we deal with that? I'm not sure I see how this implementation method can being patched to fix that use case.

Reply:
IMO what you describe is basically any sort of user change that is not visual - hence there won't be any feedback, just as there is none in other apps when you make non visual changes. I don't think we have to necessarily deal with this in any way - when I'm executing undo (in any application) I expect that every single change will be reverted up until the point that I decide to stop.

If we really have to deal with the situation that you described, then I think we should separate component definition from the dom state and keep it separately

Reply:
* If I take google sheets and make edits across different sheets, upon undo I get taken to the sheet where the change is being undone.
* on google docs, when I make changes in different places in large documents, I scroll to the relevant place in the document where the change has been undone
* in figma, when I try to undo a change that happened on a different page, I get first taken to that page and need to do a second undo command to actually make the change

Reply:
Do you see any way to do this without separating them from the main dom?

Reply:
> Do you see any way to do this without separating them from the main dom?

_To answer the exact question, yes, using the [command pattern](https://en.wikipedia.org/wiki/Command_pattern) to create an undo stack of reducer actions instead of states._

But, I don't think we need to make it that complicated, it will have its own set of edge cases. So let's scratch that. When I was just looking at these 3 applications I realized that none of them have localized undo history stacks. Their undo history is global as well, which means that if we want to provide similar behavior we can still base our undo stack on dom state. (I had some wrong assumptions about how their undo works, it's simpler than I thought)
The thing that is missing in Toolpad vs. them is the visual cue. i.e. We need to detect whether an undo concerns a change that is currently in the view. And if not, we should switch to the view first.

And when combining dom updates together for the undo stack (above in the throttle), we should also detect whether two subsequent changes happen within the same view and never colocate two changes in different views. Or at least let's think about how we will implement this because it will probably share several low-level constructs. 

High level pseudo-code I'm thinking:

```tsx
function getViewForDom(dom: AppDom): ToolpadView
function isSameView(view1: ToolpadView, view2: ToolpadView)
function navigateToView(ToolpadView)
```

But I'm just including this as illustration of where my current thinking is to help illustrate. Don't take it literally, it may be based on wrong assumptions.

Reply:
> we should also detect whether two subsequent changes happen within the same view and never colocate two changes in different views

I'm wondering if this is even possible right now, I mean to execute changes in 2 different views 🤔 (from user perspective)

> The thing that is missing in Toolpad vs. them is the visual cue. i.e. We need to detect whether an undo concerns a change that is currently in the view. And if not, we should switch to the view first.

I agree, this would be a good UX improvement and IMO we should do that, I'd just prefer that it wouldn't be a blocker for the functionality itself. I imagine this could be a fairly big change that could also require to look into performance as we would probably be doing snapshot comparison on each undo/redo action for the entire list of 2 doms 🤔

Reply:
>  I imagine this could be a fairly big change that could also require to look into performance as we would probably be doing snapshot comparison on each undo/redo action for the entire list of 2 doms 🤔

Just to point out that these doms are immutable and their properties can be compared using referential equality. Two subsequent dom states will in most cases have only one node changed. We already used to diff on every change prior to the following PR: https://github.com/mui/mui-toolpad/pull/1167/files#diff-d90295f30b6ff19e7737451a20905e502d606e97be9a94d862c1a16ef8b6017dL236-L241

Reply:
Ah right, that should simplify things 👍 nevertheless, I'd like to approach this UX improvement in a follow up

Reply:
I don't mind splitting the PRs, but then let's add a todo list to https://github.com/mui/mui-toolpad/issues/226 and not close it yet.

Reply:
Done